AX: Collapse AccessibilityTreeItem into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 5950b280ef0f151334f65057f8ad3348a6cc7ad4
<pre>
AX: Collapse AccessibilityTreeItem into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=298459">https://bugs.webkit.org/show_bug.cgi?id=298459</a>
<a href="https://rdar.apple.com/159953769">rdar://159953769</a>

Reviewed by Tyler Wilcock and Chris Fleizach.

Collapses all of the logic from AccessibilityTreeItem into AccessibilityRenderObject,
removing the former class completely.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::isAccessibilityTreeItem): Deleted.
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::isTreeItemValid const):
(WebCore::AXObjectRareData::setIsTreeItemValid):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::supportsCheckedState const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInlines.h:
(WebCore::AccessibilityObject::hasTreeItemRole const):
(WebCore::AccessibilityObject::supportsCheckedState const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::containingTree const):
(WebCore::AccessibilityRenderObject::shouldIgnoreAttributeRole const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTree.cpp:
* Source/WebCore/accessibility/AccessibilityTreeItem.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTreeItem.h: Removed.

Canonical link: <a href="https://commits.webkit.org/299650@main">https://commits.webkit.org/299650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df70e14cfadbe955764a1a8c82154db0ce2498d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30028 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71756 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d95a1c81-88fa-4a0a-bb36-8bf4d5aa7b16) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40072 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47954 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90924 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0179ed44-5f5d-43eb-b39d-93524855ecad) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71432 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f9c2fa95-17aa-4f1b-a5b0-0193660dde6f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31024 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25420 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69616 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25612 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128930 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46604 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35307 "Found 1 new test failure: http/tests/cache/disk-cache/disk-cache-302-status-code.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99503 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46969 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99346 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25233 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44776 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22795 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46466 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52172 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45932 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49281 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47618 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AccessibilityObjectInlines.h:
```diff
@@ -65,6 +65,12 @@ inline bool AccessibilityObject::isNonNativeTextControl() const
 
+inline bool AccessibilityObject::hasTreeItemRole() const
+{
+    Element* element = this->element();
```
are we preferring Element * or auto * in these cases?

hoffmanjoshua replied:
We actually prefer `RefPtr` oops 😅 , I'll swap that in here

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -470,6 +470,8 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
 
     virtual std::optional<InputType::Type> inputType() const = 0;
 
+    virtual bool hasTreeItemRole() const = 0;
```
Why is this virtual on AXCoreObject and not on AccessibilityObject?

hoffmanjoshua replied:
Yeah this is unnecessary, we can make it non-virtual on AccessibilityObject

twilco commented on Source/WebCore/accessibility/AXObjectRareData.h:
```diff
@@ -78,6 +82,11 @@ class AXObjectRareData : public CanMakeCheckedPtr<AXObjectRareData> {
     // End table-related fields.
+
+    // Begin tree-grid-item related fields.
+    bool m_isTreeItemValid;
```
Kind of wonder if a name like `m_isValidTreeItem` makes more sense? The old name made fine sense in the context of existing in a TreeItem subclass, but `isTreeItemValid` feels like it reads backwards when placed into this rare data class.

hoffmanjoshua replied:
updated!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2286,8 +2293,11 @@ AccessibilityObject* AccessibilityRenderObject::observableObject() const
+bool AccessibilityRenderObject::shouldIgnoreAttributeRole()
 {
+    if (hasTreeItemRole())
+        return hasRareData() && !rareData()->isTreeItemValid();
```
Why can this no longer be const?

hoffmanjoshua replied:
I used to use `ensureRareData` here, which is why I made it non-const. Since I'm not doing that anymore, I restored  it to be `const`

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2300,6 +2310,11 @@ AccessibilityRole AccessibilityRenderObject::determineAccessibilityRole()
 
+    if (hasTreeItemRole()) {
+        CheckedRef rareData = ensureRareData();
+        rareData->setIsTreeItemValid(containingTree());
```
Why have the local variable? Would ensureRareData().setIsTreeItemValid(containingTree()) not work?

hoffmanjoshua replied:
Fixed

twilco commented on Source/WebCore/accessibility/AccessibilityObjectInlines.h:
```diff
@@ -65,6 +65,12 @@ inline bool AccessibilityObject::isNonNativeTextControl() const
+inline bool AccessibilityObject::hasTreeItemRole() const
+{
+    RefPtr element = this->element();
+    return element ? hasRole(*element, "treeitem"_s) : false;
```
Can be written more simply as:

return element && hasRole(*element, "treeitem"_s);

hoffmanjoshua committed:
AX: Collapse AccessibilityTreeItem into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=298459
rdar://159953769

Reviewed by Tyler Wilcock and Chris Fleizach.

Collapses all of the logic from AccessibilityTreeItem into AccessibilityRenderObject,
removing the former class completely.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::isAccessibilityTreeItem): Deleted.
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::isTreeItemValid const):
(WebCore::AXObjectRareData::setIsTreeItemValid):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::supportsCheckedState const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInlines.h:
(WebCore::AccessibilityObject::hasTreeItemRole const):
(WebCore::AccessibilityObject::supportsCheckedState const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::containingTree const):
(WebCore::AccessibilityRenderObject::shouldIgnoreAttributeRole const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTree.cpp:
* Source/WebCore/accessibility/AccessibilityTreeItem.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTreeItem.h: Removed.

Canonical link: https://commits.webkit.org/299650@main
